core/src/main/java/org/apache/iceberg/avro/PruneColumns.java

```diff
@@ -100,36 +116,48 @@ public Schema union(Schema union, List<Schema> options) {
   }
 
   @Override
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
   public Schema array(Schema array, Schema element) {
-    if (array.getLogicalType() instanceof LogicalMap) {
+    if (array.getLogicalType() instanceof LogicalMap || AvroSchemaUtil.isKeyValueSchema(array.getElementType())) {
       Schema keyValue = array.getElementType();
-      int keyId = AvroSchemaUtil.getFieldId(keyValue.getField("key"));
-      int valueId = AvroSchemaUtil.getFieldId(keyValue.getField("value"));
+      Integer keyId = fieldId(keyValue.getField("key"));
+      Integer valueId = fieldId(keyValue.getField("value"));
+      if (keyId == null) {
+        if (valueId != null) {
+          LOG.warn("Map schema {} has value id but not key id", array);
+        }
+        return null;
+      }
 
       // if either key or value is selected, the whole map must be projected
       if (selectedIds.contains(keyId) || selectedIds.contains(valueId)) {
-        return array;
+        return complexMapWithIds(array, keyId, valueId);
       } else if (element != null) {
-        if (keyValue.getField("value").schema() != element.getField("value").schema()) {
+        if (keyValue.getField("key").schema() != element.getField("key").schema() ||
+            keyValue.getField("value").schema() != element.getField("value").schema()) {
           // the value must be a projection
           return AvroSchemaUtil.createMap(
-              keyId, keyValue.getField("key").schema(),
+              keyId, element.getField("key").schema(),
```

Comment:
This should continue to use the schema from `keyValue` because map keys can't be partially projected. They must be fully projected so we always use the original key type.

Reply:
Thanks, I'll update

Reply:
On second thought,  we'd have to use `element.getField("key").schema(),` as the key [if it is a complex type] could contain newly assigned field ids  which the original key did not have.